File: src/test/cspom/extension/ExtensionConstraintTest.java
package cspom.extension;


import java.util.Arrays;

import org.junit.Before;

import cspom.extension.Extension;
import cspom.extension.ExtensionConstraint;
import cspom.variable.Domain;
import cspom.variable.ExtensiveDomain;
import cspom.variable.Variable;

public class ExtensionConstraintTest {

	@Before
	public void setUp() throws Exception {
	}


	public static void main(String[] args) {
		Extension relation = new Extension("test", 3, 4, true, new Number[][] {
				{ 1, 2, 3 }, { 2, 3, 4 }, { 3, 4, 5 }, { 4, 5, 6 } });

		Domain domain = new ExtensiveDomain("d1", Arrays.asList(new Number[] {
				1, 2, 3, 4, 5, 6 }));

		Variable v1 = new Variable("v1", domain);
		Variable v2 = new Variable("v2", domain);
		Variable v3 = new Variable("v3", domain);

		ExtensionConstraint constraint = new ExtensionConstraint("c1", Arrays
				.asList(new Variable[] { v1, v2, v3 }), relation);

		System.out.println(constraint);
		System.out.println(constraint.getRelation().tupleString());

		ExtensionConstraint c2 = constraint.standardize(Arrays
				.asList(new Variable[] { v2, v3, v1 }));

		System.out.println(c2);
		System.out.println(c2.getRelation().tupleString());

	}
}

File: src/test/scala/cspom/extension/ExtensionConstraintTest.scala
package cspom.extension;

import org.junit.Test
import org.junit.Before
import org.junit.Assert._
import org.hamcrest.CoreMatchers._
import org.junit.matchers.JUnitMatchers._

final class ExtensionConstraintTest {

  var relation: Trie = null;

  @Before
  def setUp() {
    relation = Trie.empty;
    relation += (2, 5, 5);
    relation += (3, 5, 5);
  }

  @Test
  def testEvaluate() {
    assertFalse(relation.contains(1, 2, 3));
    assertTrue(relation.contains(2, 5, 5));
  }

  @Test
  def testTuplesToString() {
    assertThat(relation.tupleString, either(equalTo("2 5 5|3 5 5")).or(equalTo("3 5 5|2 5 5")));
  }

  @Test
  def testSize() {
    assertEquals(2, relation.size)
    assertEquals(3, relation.depth)
  }

  @Test
  def testAddTuple() {
    relation += (2, 5, 5);
    assertEquals(2, relation.size);
  }

  @Test
  def testRemove() {
    relation -= Array(2, 5, 5)
    assertEquals(1, relation.size)
    assertEquals("3 5 5", relation.tupleString)
    relation -= Array(3, 5, 5)
    assertEquals(0, relation.size)
    assertEquals("", relation.tupleString)
  }

  @Test
  def testFilter() {
    assertSame(Trie.empty, relation.filterTrie({
      case (depth, value) => depth != 1 || value != 5
    }))
    assertEquals("2 5 5", relation.filterTrie({
      case (depth, value) => depth != 0 || value != 3
    }).tupleString)
  }
}
